superset/utils/excel.py

```diff
@@ -19,16 +19,36 @@
 
 import pandas as pd
 
+from superset.utils.core import GenericDataType
 
-def df_to_excel(df: pd.DataFrame, **kwargs: Any) -> Any:
+
+def df_to_excel(
+    df: pd.DataFrame, coltypes: list[GenericDataType], **kwargs: Any
+) -> Any:
     output = io.BytesIO()
 
     # timezones are not supported
     for column in df.select_dtypes(include=["datetimetz"]).columns:
         df[column] = df[column].astype(str)
 
+    ndf = copy_df_with_datatype_conversion(df, coltypes)
```

Comment:
@squalou Can the column types be extracted from the `DataFrame` like we are doing for timezones

```
for column in df.select_dtypes(include=["datetimetz"]).columns:
        df[column] = df[column].astype(str)
```

or are they not matching what's defined in `coltypes`?

Reply:
Thing is : you could have data with leading zeros for instance, defined as string in the original datasource, and you want them as string in the end, or else excel would omit leading zeros, maybe round things, whatever.

I don't know much about DataFrame object, and the whole issue I'm trying to solve is to get the datatype *as it was defined* as the column type to begin with.

If you known of a way to get this data from the DataFrame itself, it sure would be MUCH more elegant than forwarding the column_types all way down to excel export.

To be honest, I discovered panda objects only to make up this patch  ... so if you think this metadata can be stored / handled differently, I'll be glad to learn.

I also did not want to take any risk of changing the original dataframe object as I have no idea of global superset architecture.

Reply:
I almost forgot the most important part : I have absolutely no idea how to know what dtypes are 'stored' in the DataFrame at the beginning :) so I used the 'coltypes' that seemed to be tehre for a reason :)